extras/ai-battle/HeadlessGame.cpp

```diff
@@ -0,0 +1,264 @@
+            for(unsigned playerId = 0; playerId < world_.GetNumPlayers(); ++playerId)
+            {
+                world_.GetPlayer(playerId);
+                AIPlayer* player = players_[playerId].get();
+                PlayerGameCommands cmds;
+                cmds.gcs = player->FetchGameCommands();
+
+                if(replay_.IsRecording() && !cmds.gcs.empty())
+                {
+                    cmds.checksum = checksum;
+                    replay_.AddGameCommand(em_.GetCurrentGF(), playerId, cmds);
+                }
+
+                for(const gc::GameCommandPtr& gc : cmds.gcs)
+                    gc->Execute(world_, player->GetPlayerId());
+            }
+        }
+
+        for(auto& player : players_)
+            player->RunGF(em_.GetCurrentGF(), isnfw);
+
+        game_.RunGF();
+
+        if(replay_.IsRecording())
+            replay_.UpdateLastGF(em_.GetCurrentGF());
+
+        if(std::chrono::steady_clock::now() > nextReport)
+        {
+            nextReport += std::chrono::seconds(1);
+            PrintState();
+        }
+    }
+    PrintState();
+}
+
+void HeadlessGame::Close()
+{
+    bnw::cout << '\n';
+
+    if(replay_.IsRecording())
+    {
+        replay_.StopRecording();
+        bnw::cout << "Replay written to " << canonical(replayPath_) << '\n';
+    }
+
+    replay_.Close();
+}
+
+void HeadlessGame::StartReplay(const bfs::path& path, unsigned random_init)
+{
+    // Remove old replay
+    bfs::remove(path);
+
+    replayPath_ = path;
+
+    MapInfo mapInfo;
+    mapInfo.filepath = map_;
+    mapInfo.mapData.CompressFromFile(mapInfo.filepath, &mapInfo.mapChecksum);
+    mapInfo.type = MapType::OldMap;
+
+    replay_.random_init = random_init;
+    for(unsigned playerId = 0; playerId < world_.GetNumPlayers(); ++playerId)
+        replay_.AddPlayer(world_.GetPlayer(playerId));
+    replay_.ggs = game_.ggs_;
+    if(!replay_.StartRecording(path, mapInfo))
+        throw std::runtime_error("Replayfile could not be opened!");
+}
+
+void HeadlessGame::SaveGame(const bfs::path& path) const
+{
+    // Remove old savegame
+    bfs::remove(path);
+
+    Savegame save;
+    for(unsigned playerId = 0; playerId < world_.GetNumPlayers(); ++playerId)
+        save.AddPlayer(world_.GetPlayer(playerId));
+    save.ggs = game_.ggs_;
+    save.ggs.exploration = Exploration::Disabled; // no FOW
+    save.start_gf = em_.GetCurrentGF();
+    save.sgd.MakeSnapshot(game_);
+    save.Save(path, "AI Battle");
+
+    bnw::cout << "Savegame written to " << canonical(path) << '\n';
+}
+
+std::string ToString(const std::chrono::milliseconds& time)
+{
+    char buffer[90];
+    auto hours = std::chrono::duration_cast<std::chrono::hours>(time);
+    auto minutes = std::chrono::duration_cast<std::chrono::minutes>(time % std::chrono::hours(1));
+    auto seconds = std::chrono::duration_cast<std::chrono::seconds>(time % std::chrono::minutes(1));
+    snprintf(buffer, 90, "%03ld:%02ld:%02ld", static_cast<long int>(hours.count()),
```

Comment:
No magic numbers if possible:
```suggestion
    snprintf(buffer, std::size(buffer), "%03ld:%02ld:%02ld", static_cast<long int>(hours.count()),
```